### ISSUE TITLE
best practice for saving / restoring the position in the dataset (iterator)

### ISSUE THREAD

**Issue by hrbigelow** [google/orbax#413]
Hi,

I noticed [here](https://github.com/google/orbax/blob/main/docs/checkpoint.md#checkpointmanager) it's mentioned that a `tf.data.Iterator` could be checkpointed.  

I tried this, but it didn't work:

```python
import orbax.checkpoint
checkpointer = orbax.checkpoint.PyTreeCheckpointer()

ds = tf.data.Dataset.load(dataset_path)
ds = ds.shuffle(buffer_size=10000, seed=12345)
it = iter(ds)

# ... train for some number of iterations
checkpointer.save(checkpoint_path, { 'ds_iterator': it })
```

and then to restore I tried:

```python
restored = checkpointer.restore(checkpoint_path)
print(restored)
# { 'ds_iterator': None }
```

In particular, how would the checkpointer "know" that when restoring, it needs to re-do the shuffle of the dataset, and with the particular random seed?  Seems like none of that information could be held in the iterator, so this must not be the right way.  Any advice appreciated!

Thanks,

Henry

**Comment by cpgaffney1** [comment]
Yes, this should be pretty easy to do, but you need to implement your own CheckpointHandler (see [docs](https://orbax.readthedocs.io/en/latest/api_reference/checkpoint.checkpoint_handlers.html#checkpointhandler)). The object you're currently using the save and restore is only built for PyTrees. 

Orbax doesn't provide a default implementation for `tf.data.Iterator` because we don't want to include a dependency on Tensorflow. As a result, users are expected to implement it, since it's a somewhat more specialized use case. It should be pretty straightforward to do so - feel free to take a look [here](https://github.com/google-research/t5x/blob/main/t5x/checkpoints.py#L1889) for some inspiration.

**Comment by hrbigelow** [comment]
Excellent.  Thanks @cpgaffney1.  And it does make sense to keep it external to orbax.
